### PR TITLE
Calculate the build cache key when the build scan plugin has been applied

### PR DESCRIPTION
**Pull request by wolfs**

### CODE REVIEW

**Comment by ldaley** [comment]
Related to https://github.com/gradle/build-cache/issues/458

File: subprojects/base-services/src/main/java/org/gradle/internal/hash/HashCode.java [hunk 1]
```diff
@@ -96,6 +96,10 @@ public int length() {
         return bytes.clone();
     }
 
+    public void appendToHasher(Hasher hasher) {
+        hasher.putBytes(bytes);
+    }
+
```
Comment (lptr) [inline subprojects/base-services/src/main/java/org/gradle/internal/hash/HashCode.java]: I'd prefer if we had this in the form of `Hasher.putHashCode(HashCode)` and `MessageDigestHasher` would have the ability to access the raw `bytes` inside the `HashCode` via a package-private accessor.

File: subprojects/core/src/integTest/groovy/org/gradle/api/internal/tasks/SnapshotTaskInputsOperationIntegrationTest.groovy [hunk 2]
```diff
@@ -46,6 +48,40 @@ class SnapshotTaskInputsOperationIntegrationTest extends AbstractIntegrationSpec
...
+        then:
+        def result = operations.first(SnapshotTaskInputsBuildOperationType).result
+
+        then:
+        result.buildCacheKey != null
+        result.inputHashes.keySet() == ['input1', 'input2'] as Set
+        result.outputPropertyNames == ['outputFile1', 'outputFile2']
+    }
+
     def "task output caching key is not exposed when build cache is disabled"() {
```
Comment (lptr) [inline subprojects/core/src/integTest/groovy/org/gradle/api/internal/tasks/SnapshotTaskInputsOperationIntegrationTest.groovy]: Should we update the name of this test, too?

File: subprojects/core/src/integTest/groovy/org/gradle/api/internal/tasks/SnapshotTaskInputsOperationIntegrationTest.groovy [hunk 3]
```diff
@@ -46,6 +48,40 @@ class SnapshotTaskInputsOperationIntegrationTest extends AbstractIntegrationSpec
...
+                }
+            }
+            
+            apply plugin: "com.gradle.build-scan"
+            buildScan {
+                licenseAgreementUrl = 'https://gradle.com/terms-of-service'
+                licenseAgree = 'yes'
+            }
+        """.stripIndent()
+        // Using the embedded executer, we get an NPE. We force forking by requiring a Gradle distribution
```
Comment (lptr) [inline subprojects/core/src/integTest/groovy/org/gradle/api/internal/tasks/SnapshotTaskInputsOperationIntegrationTest.groovy]: What throws the NPE?

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/package-info.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+@NonNullApi
+package org.gradle.api.tasks;
+
+import org.gradle.api.NonNullApi;
```
Comment (lptr) [inline subprojects/core/src/integTest/groovy/org/gradle/api/tasks/package-info.java]: This changes a lot of things, and will need a thorough review. I think it would be best to do it in a separate PR.

File: subprojects/core/src/integTest/groovy/org/gradle/internal/scan/config/BuildScanConfigIntegrationTest.groovy [hunk 5]
```diff
@@ -36,9 +37,14 @@ Please see https://gradle.com/scans/help/gradle-incompatible-plugin-version for
                     def c = services.get(${BuildScanConfigProvider.name}).collect([getVersion: { "$pluginVersionNumber" }] as $BuildScanPluginMetadata.name) 
                     println "buildScan.enabled: " + c.enabled 
                     println "buildScan.disabled: " + c.disabled 
-                """
+                """.stripIndent()
```
Comment (lptr) [inline subprojects/core/src/integTest/groovy/org/gradle/internal/scan/config/BuildScanConfigIntegrationTest.groovy]: Why is adding `stripIndent()` useful? It looks like unnecessary noise to me.

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/DefaultFileCollectionSnapshot.java [hunk 6]
```diff
@@ -79,12 +100,23 @@ public boolean isEmpty() {
...
+        }
         return compareStrategy.iterateContentChangesSince(snapshots, oldSnapshot.getSnapshots(), fileType, pathIsAbsolute, includeAdded);
     }
 
     @Override
+    public HashCode getHash() {
+        if (hashCode == null) {
+            hashCode = hashCodeSupplier.get();
+        }
+        return hashCode;
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/DefaultFileCollectionSnapshot.java]: What is the use of the memoizing `Supplier` if we also do a `null`-check here? If you want to use a `Supplier`, you could wrap the `hashCode` in a constant one if it's provided in the constructor, or use a calculating one if it's not. We could use our own `Factory`, too.

But I would simply write the code out here. I don't think it's worth creating all these objects just to hide that small amount of code.

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/DefaultFileCollectionSnapshot.java [hunk 7]
```diff
@@ -122,20 +154,29 @@ public String toString() {
...
+            HashCode hash = hasHash ? hashCodeSerializer.read(decoder) : null;
             Map<String, NormalizedFileSnapshot> snapshots = snapshotMapSerializer.read(decoder);
             boolean pathIsUnique = decoder.readBoolean();
-            return new DefaultFileCollectionSnapshot(snapshots, compareStrategy, pathIsUnique);
+            return new DefaultFileCollectionSnapshot(snapshots, hash, compareStrategy, pathIsUnique);
         }
 
         public void write(Encoder encoder, DefaultFileCollectionSnapshot value) throws Exception {
             encoder.writeSmallInt(value.compareStrategy.ordinal());
+            boolean hasHash = value.getHash() != null;
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/DefaultFileCollectionSnapshot.java]: Shouldn't we serialize `value.hashCode` here instead?

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/FileCollectionSnapshot.java [hunk 8]
```diff
@@ -30,7 +32,7 @@
  * An immutable snapshot of some aspects of the contents and meta-data of a collection of files or directories.
  */
 public interface FileCollectionSnapshot extends Snapshot {
-    FileCollectionSnapshot EMPTY = new DefaultFileCollectionSnapshot(Collections.<String, NormalizedFileSnapshot>emptyMap(), UNORDERED, true);
+    FileCollectionSnapshot EMPTY = new DefaultFileCollectionSnapshot(Collections.<String, NormalizedFileSnapshot>emptyMap(), new DefaultBuildCacheHasher().hash(), UNORDERED, true);
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/FileCollectionSnapshot.java]: Use `HashCode.fromInt(1234)` or something instead of instantiating a hasher.

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/FileCollectionSnapshot.java [hunk 9]
```diff
@@ -39,6 +41,8 @@
      */
     Iterator<TaskStateChange> iterateContentChangesSince(FileCollectionSnapshot oldSnapshot, String title, boolean includeAdded);
 
+    HashCode getHash();
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/FileCollectionSnapshot.java]: Would be nice to have some Javadoc, given that the other guys also have some.

Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/FileCollectionSnapshot.java]: We could even mention that we short-circuit the comparison if we already have a hash code.

File: subprojects/core/src/integTest/groovy/org/gradle/internal/scan/config/BuildScanConfigIntegrationTest.groovy [hunk 10]
```diff
@@ -36,9 +37,14 @@ Please see https://gradle.com/scans/help/gradle-incompatible-plugin-version for
...
             }
 
+            buildFile << """    
+                def pluginApplied = services.get(${BuildScanPluginApplied.name}).isBuildScanPluginApplied()
+                println "buildScan plugin applied: " + pluginApplied
+            """.stripIndent()
```
Comment (lptr) [inline subprojects/core/src/integTest/groovy/org/gradle/internal/scan/config/BuildScanConfigIntegrationTest.groovy]: Is there no other way to tell if the build scan plugin was applied? Like, check if we have it among the applied plugins or something? Adding a production type just for this seems unnecessarily complicated.

File: subprojects/core/src/main/java/org/gradle/internal/scan/config/package-info.java [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+@NonNullApi
+package org.gradle.internal.scan.config;
+
+import org.gradle.api.NonNullApi;
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/internal/scan/config/package-info.java]: Have you checked all the methods for nullability problems? Even so I would add this as a separate PR.

File: subprojects/base-services/src/main/java/org/gradle/internal/hash/HashCode.java (hunk 1 above)
Comment (lptr) [reply subprojects/base-services/src/main/java/org/gradle/internal/hash/HashCode.java]: If we did that we could even merge the `Hasher` and `BuildCacheHasher` interfaces. We don't necessarily have to do that part now, though.

File: subprojects/core/src/integTest/groovy/org/gradle/internal/scan/config/BuildScanConfigIntegrationTest.groovy (hunk 10 above)
Comment (wolfs) [reply subprojects/core/src/integTest/groovy/org/gradle/internal/scan/config/BuildScanConfigIntegrationTest.groovy]: I discussed with @ldaley about it. I had the check for the applied plugin before. I guess checking the flag is faster then checking if the plugin has been applied to the root project. It should also be more future proof, given that we have a problem in getting the right root project when running with composite builds or in the buildSrc project.

File: subprojects/core/src/integTest/groovy/org/gradle/api/internal/tasks/SnapshotTaskInputsOperationIntegrationTest.groovy (hunk 3 above)
Comment (wolfs) [reply subprojects/core/src/integTest/groovy/org/gradle/api/internal/tasks/SnapshotTaskInputsOperationIntegrationTest.groovy]: I added a better explanation.

File: subprojects/core/src/integTest/groovy/org/gradle/api/internal/tasks/SnapshotTaskInputsOperationIntegrationTest.groovy (hunk 2 above)
Comment (wolfs) [reply subprojects/core/src/integTest/groovy/org/gradle/api/internal/tasks/SnapshotTaskInputsOperationIntegrationTest.groovy]: Changed the name.

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/package-info.java (hunk 4 above)
Comment (wolfs) [reply subprojects/core/src/integTest/groovy/org/gradle/api/tasks/package-info.java]: I removed both `package-info.java` files.

File: subprojects/core/src/integTest/groovy/org/gradle/internal/scan/config/BuildScanConfigIntegrationTest.groovy (hunk 5 above)
Comment (wolfs) [reply subprojects/core/src/integTest/groovy/org/gradle/internal/scan/config/BuildScanConfigIntegrationTest.groovy]: I removed it. I actually like adding `stripIdent` when I write integration tests, since then generated build files look nicer and sometimes I look at them. You prefer not using `stripIdent()` at all in integration tests?

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/DefaultFileCollectionSnapshot.java (hunk 6 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/DefaultFileCollectionSnapshot.java]: I removed the memoizing supplier and inlined the code.

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/DefaultFileCollectionSnapshot.java (hunk 7 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/DefaultFileCollectionSnapshot.java]: Good catch. That's where the performance regressions are coming from I guess. I have no clue why I didn't see this in the profiler snapshots I look at. :(

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/FileCollectionSnapshot.java (hunk 8 above)
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/FileCollectionSnapshot.java]: Sorry, thought this was test code. Please carry on.

File: subprojects/base-services/src/main/java/org/gradle/internal/hash/HashCode.java (hunk 1 above)
Comment (wolfs) [reply subprojects/base-services/src/main/java/org/gradle/internal/hash/HashCode.java]: fixed. I didn't unify the two interfaces yet.

**Comment by wolfs** [comment]
@lptr I addressed your comments. Could you have another look?

File: subprojects/core/src/integTest/groovy/org/gradle/internal/scan/config/BuildScanConfigIntegrationTest.groovy (hunk 5 above)
Comment (lptr) [reply subprojects/core/src/integTest/groovy/org/gradle/internal/scan/config/BuildScanConfigIntegrationTest.groovy]: I tend to look far more frequently at the code of the integration test than at the actual build file. The difference in the build file is also very small: when you don't use `stripIndent()` the lines start a little to the right, so it's not even harder to read or anything. Therefore I think we should prefer cleaner integration tests over cleaner build files, and not to use `stripIndent()` without a stronger reason.

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/FileCollectionSnapshot.java [hunk 12]
```diff
@@ -41,6 +41,11 @@
      */
     Iterator<TaskStateChange> iterateContentChangesSince(FileCollectionSnapshot oldSnapshot, String title, boolean includeAdded);
 
+    /**
+     * Returns the combined hash of the contents.
+     *
+     * We short-circuit {@link #iterateContentChangesSince(FileCollectionSnapshot, String, boolean)} if we already know the hash.
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/FileCollectionSnapshot.java]: We only short circuit if we know the hash _and_ it's equal to the hash of the file collection we compare it to.

**Comment by lptr** [comment]
The changes look good. Can you add some explanation about why we do the flag thing with the build scan plugin instead of checking amongst the plugins the usual way? Otherwise I think we can merge.

**Comment by wolfs** [comment]
@ldaley Can you have a look?
